### ISSUE TITLE
event_match key in push rule conditions does not cater for field names with dots

### ISSUE THREAD

**Issue by richvdh** [matrix-org/matrix-doc#2641]
https://matrix.org/docs/spec/client_server/r0.6.1#conditions says: `key` The dot-separated field of the event to match, e.g. `content.body`. Various standard event fields (notably those in https://github.com/matrix-org/matrix-doc/pull/1849) have dots in their names. It's unclear how these are handled.

**Comment by richvdh** [comment]
In practice, it looks like `{"key": "m.foo", pattern: "bar"}` will match both:

```json
{
    "m": { "foo": "bar" }
}
```

and

```json
{
    "m.foo": "bar"
}
```

**Comment by richvdh** [comment]
related: #2637

**Comment by t3chguy** [comment]
What will it match in the case of ambiguity?

```json
{
    "m": { "foo": "bar" },
    "m.foo": "baz"
}
```

**Comment by richvdh** [comment]
IT DEPENDS :tada: 

depends what order python decides to go through the dictionary, which is probably ultimately determined by the order they arrive over the wire from the client.

so that's fun.

**Comment by jplatte** [comment]
Could we just deprecate the "dot-separated" in 

> `key` The dot-separated field of the event to match, e.g. `content.body`

?

**Comment by t3chguy** [comment]
How would you refer to match `content.body` then?

**Comment by jplatte** [comment]
Well, like in your example,

```json
{
    "content": { "body": "foo" }
}
```

**Comment by t3chguy** [comment]
My example is of an event body, how would you refer to it in the `key` string for the push rule which is what this issue is about?

**Comment by jplatte** [comment]
Okay, I misunderstood then. In that case, and if we want to continue only matching on one key (which is probably a good idea, otherwise we'd open another can of worms), maybe we could adopt [JSON pointers](https://tools.ietf.org/html/rfc6901)? That would make the implementation much more straight-forward, though we'd have to figure out the compatibility story for that.

**Comment by richvdh** [comment]
yeah the problem is that we *could* do any number of better things, but it's going to be hard to maintain compatibility. Ultimately servers are probably going to have to support either of two formats. Meanwhile it's more of a wart than a practical problem.

**Comment by uhoreg** [comment]
FWIW, there's a similar issue with filters, and it's solved there by escaping "." in field names by a "\\".
![image](https://user-images.githubusercontent.com/1012976/84928892-195b7680-b09d-11ea-839b-fa2bbf1193d3.png)

**Comment by schmop** [comment]
Escaping would be fairly cheap to implement and it would fix the issues with dots in a prop key.
Would love to see this!
